src/mesh/abstract_tree.jl

```diff
@@ -1,201 +1,62 @@
-Tree(::Val{NDIMS}, args...) where NDIMS = Tree{NDIMS}(args...)
-
-# Create and initialize tree
-function Tree{NDIMS}(capacity::Int, center::AbstractArray{Float64},
-                 length::Real, periodicity=true) where NDIMS
-  # Create instance
-  t = Tree{NDIMS}(capacity)
-
-  # Initialize root cell
-  init!(t, center, length, periodicity)
-
-  return t
-end
-
-# Constructor accepting a single number as center (as opposed to an array) for 1D
-Tree{1}(cap::Int, center::Real, len::Real, periodicity=true) = Tree{1}(cap, [convert(Float64, center)], len, periodicity)
-
-
-# Clear tree with deleting data structures, store center and length, and create root cell
-function init!(t::Tree, center::AbstractArray{Float64}, length::Real, periodicity=true)
-  clear!(t)
-
-  # Set domain information
-  t.center_level_0 = center
-  t.length_level_0 = length
-
-  # Create root cell
-  t.length += 1
-  t.parent_ids[1] = 0
-  t.child_ids[:, 1] .= 0
-  t.levels[1] = 0
-  t.coordinates[:, 1] .= t.center_level_0
-  t.original_cell_ids[1] = 0
-
-  # Set neighbor ids: for each periodic direction, the level-0 cell is its own neighbor
-  if all(periodicity)
-    # Also catches case where periodicity = true
-    t.neighbor_ids[:, 1] .= 1
-    t.periodicity = ntuple(x->true, ndims(t))
-  elseif !any(periodicity)
-    # Also catches case where periodicity = false
-    t.neighbor_ids[:, 1] .= 0
-    t.periodicity = ntuple(x->false, ndims(t))
-  else
-    # Default case if periodicity is an iterable
-    for dimension in 1:ndims(t)
-      if periodicity[dimension]
-        t.neighbor_ids[2 * dimension - 1, 1] = 1
-        t.neighbor_ids[2 * dimension - 0, 1] = 1
-      else
-        t.neighbor_ids[2 * dimension - 1, 1] = 0
-        t.neighbor_ids[2 * dimension - 0, 1] = 0
-      end
-    end
-
-    t.periodicity = Tuple(periodicity)
-  end
-end
-
-
-# Convenience output for debugging
-function Base.show(io::IO, t::Tree{NDIMS}) where NDIMS
-  l = t.length
-  println(io, '*'^20)
-  println(io, "t.parent_ids[1:l] = $(t.parent_ids[1:l])")
-  println(io, "transpose(t.child_ids[:, 1:l]) = $(transpose(t.child_ids[:, 1:l]))")
-  println(io, "transpose(t.neighbor_ids[:, 1:l]) = $(transpose(t.neighbor_ids[:, 1:l]))")
-  println(io, "t.levels[1:l] = $(t.levels[1:l])")
-  println(io, "transpose(t.coordinates[:, 1:l]) = $(transpose(t.coordinates[:, 1:l]))")
-  println(io, "t.original_cell_ids[1:l] = $(t.original_cell_ids[1:l])")
-  println(io, "t.capacity = $(t.capacity)")
-  println(io, "t.length = $(t.length)")
-  println(io, "t.dummy = $(t.dummy)")
-  println(io, "t.center_level_0 = $(t.center_level_0)")
-  println(io, "t.length_level_0 = $(t.length_level_0)")
-  println(io, '*'^20)
-end
+abstract type AbstractTree{NDIMS} <: AbstractContainer end
 
 # Type traits to obtain dimension
-@inline Base.ndims(t::Type{Tree{NDIMS}}) where NDIMS = NDIMS
-@inline Base.ndims(t::Tree) = ndims(typeof(t))
+@inline Base.ndims(::Type{AbstractTree{NDIMS}}) where NDIMS = NDIMS
```

Comment:
Do we use this somewhere? Usually, it's considered best practice to use instances and not a type.

Reply:
No, you're right. Fixed in b84fe79.